backends/dpdk/dpdkProgram.cpp

```
@@ -0,0 +1,282 @@
+  for (auto state : p->states) {
+    if (state->name == "start")
+      stack.push_back(*state);
+    degree_map.insert({state->name.toString(), 0});
+    state_map.insert({state->name.toString(), *state});
+  }
+  for (auto state : p->states) {
+    if (state->selectExpression) {
+      if (state->selectExpression->is<IR::SelectExpression>()) {
+        auto select = state->selectExpression->to<IR::SelectExpression>();
+        for (auto pair : select->selectCases) {
+          auto got = degree_map.find(pair->state->path->name);
+          if (got != degree_map.end()) {
+            got->second++;
+          }
+        }
+      } else if (auto path =
+                     state->selectExpression->to<IR::PathExpression>()) {
+        auto got = degree_map.find(path->path->name);
+        if (got != degree_map.end()) {
+          got->second++;
+        }
+      }
+    }
+  }
+  degree_map.erase("start");
+  state_map.erase("start");
+
+  while (stack.size() > 0) {
+    auto state = stack.back();
+    stack.pop_back();
+
+    // the main body
+    auto i = new IR::DpdkLabelStatement("L_" + state.name.toString());
+    add_instr(i);
+    auto c = state.components;
+    for (auto stat : c) {
+      DPDK::ConvertStatementToDpdk h(refmap, typemap, 0, this->collector,
+                                     csum_map);
+      stat->apply(h);
+      for (auto i : h.get_instr())
+        add_instr(i);
+    }
+    if (state.selectExpression) {
+      if (auto e = state.selectExpression->to<IR::SelectExpression>()) {
+        const IR::Expression *switch_var;
+        switch_var = e->select->components[0];
+        for (auto v : e->selectCases) {
+          if (!v->keyset->is<IR::DefaultExpression>()) {
+            add_instr(new IR::DpdkJmpEqualStatement("L_" + v->state->path->name,
+                                                    v->keyset, switch_var));
+          } else {
+            auto i = new IR::DpdkJmpStatement("L_" + v->state->path->name);
```

You are not looking at the select expression at all here. where is this done?

---

@gordonwucn please take a look.

---

`switch_var` is the SelectExpression. This is the statement that compares if SelectExpression equals one of the keys. `new IR::DpdkJmpEqualStatement("L_" + v->state->path->name, v->keyset, switch_var)`

---

But you are only looking at component[0]. You may want to give an error if there are more components?

---

Thanks for mentioning! @hanw Could you throw an error when `e->select->components` is longer than 1?

---

added.